### PR TITLE
Rake task improvements

### PR DESCRIPTION
**Pull request by Genki-S**
I made some improvements for rake task handling.
@robin850 could you review this? Since some commits have short life and are overwritten by later commits, I write summary of changes in this description. However, if you prefer me to squash these commits please let me know before starting review, I'm happy to squash and tidy up the history :smile: (I feel intermediate commits might be useful in the future since I'm doing challenging changes in later commits which might not be the way to go)

Here is a summary of changes:

## Regard name passed to `Rake::TestTask`

ref: https://github.com/Genki-S/ttnt/pull/10#discussion_r32378959

If you pass name to `Rake::TestTask` like `t.name = 'my_test'`, TTNT respect it and create tasks `ttnt:my_test:anchor` and `ttnt:my_test:run`.

https://github.com/Genki-S/ttnt/commit/7acdf3163455d5c28f0d99348eb36fc5ea9c05a0

## Drop support for executables

My first approach was to use `exe/ttnt-anchor` for anchoring and `exe/ttnt` for running tests. Since rake approach seems way better and I don't think I can maintain both approaches, I drop the executable approach and focus on rake approach.

https://github.com/Genki-S/ttnt/commit/5fcc64086714d46370dda79eb617dfcb39fb594a

## :rocket:  `TTNT::TestTask` extends `Rake::TestTask` :rocket: 

This is quite challenging one!

I found `Rake::TestTask` defining rake tasks inside itself ([code](https://github.com/ruby/rake/blob/e644af3/lib/rake/testtask.rb#L96-L115)), so I stole the idea. Besides, by extending `Rake::TestTask`, I can use useful method like `ruby_opts_string` like [here](https://github.com/Genki-S/ttnt/commit/66c80fa22bd73f037b8058250b0bbd16ed6116b0#diff-0fd3d04a90ac0051bae2bd5b2c8d1716R61). It might be debatable if this is a good design, though.

Since `TTNT::TestTask` defines rake tasks, there is no need to have `TTNT::TaskHelper`. So I moved the task definition code from `TTNT::TaskHelper` to `TTNT::TestTask` and removed `TTNT::TaskHelper`.

https://github.com/Genki-S/ttnt/commit/66c80fa22bd73f037b8058250b0bbd16ed6116b0#diff-0fd3d04a90ac0051bae2bd5b2c8d1716R61

## :rocket: Run tests by `rake ttnt:TESTNAME:run` :rocket: 

This is also a challenging one!

I implemented the idea described on #1 , which is to create new `Rake::TestTask` from within `ttnt:TESTNAME:run` task and make it run selected tests.

https://github.com/Genki-S/ttnt/commit/279819827f039b507c5f541a4a6a9858a0fdd127

## `-I#{gem_root}`, again :disappointed: 

ref: https://github.com/Genki-S/ttnt/pull/6#discussion_r32372193

I'm sorry to say that I had to put the `"-I#{gem_root}"` before running `rake ttnt:test:anchor` again... I have not yet fully investigated what made things wrong, so temporarily I'm adding that dirty line.
My guess is that it's required since `system('ruby')` is invoked under non-bundler environment, but I'm not sure... I will investigate this later.

https://github.com/Genki-S/ttnt/commit/83449fed3a739c3e9078f56fbf7a343b5c4a5111

### CODE REVIEW

**Comment by robin850** [comment]
Hey Genki! It looks pretty good apart from a few points:

* Instead of defining `find_by_name`, you can just pass `self` to `define_tasks` inside `TTNT::TestTask#initialize`. Not sure if it's useful to keep `instances` then.

* > I don't think I can maintain both approaches

  Well, you could, extracting the logic of each task inside methods and calling them from the Rake tasks and the scripts but I'm not sure this is really useful. With the Rake task approach, the customization (pattern, load path, etc.) is easier and this is consistent with the existing testing tasks so +1 on this change.

* Regarding the `-I#{gem_root}` thingy, actually I should have mentioned that RubyGems monkey-patches `require` but that only works for released gem and it might not pick the right version so yes, I was wrong, we should actually keep it! :+1:

* As for the change that makes `TTNT::TestTask` extend from `Rake::TestTask`, I don't think that's a good idea. Actually for now we are only supporting this kind of task but in future we could support other testing frameworks like RSpec.

  Anyway, inheritance is a pretty bad pattern here. You should rather rely on [composition](https://en.wikipedia.org/wiki/Composition_over_inheritance) for now ; we could create an abstraction class later that would handle the support of different frameworks.

Also, a few things regarding commits. I know it's not always easy to know when to gather changes or create separate commits. At least, if you want to gather changes, make sure that they are closely related. For instance, you should not put in a single commit the extending of `TTNT::TestTask` from `Rake::TestTask` with the change that removes the `task_helper.rb` file.

Moreover, try to avoid as much as possible merge commits inside a pull request. If you are merging two branches, you can use `git rebase` to avoid that:

```
$ git checkout the-branch-to-merge
$ git rebase the-main-branch
$ git checkout the-main-branch
$ git merge the-branch-to-merge
```

If the branch is already messy, you can do:

```
$ git branch -m my-branch messy-branch
$ git checkout -b my-branch a-clean-branch

# Then select commits from `messy-branch` that
# you need on the new branch
$ git cherry-pick sha
```

It will rename the messy branch and create a new one atop a clean branch. `git cherry-pick` aims at porting some commits from a branch to another.

Also, if you want, you can have a look at [this article](https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message) for commit messages. Managing commits may be the less funny part of a patch but if it's done well, it can really make you save some time when fixing bugs or tracking regressions in the future.

Finally, you can create as many pull requests as you wish. Reviewing little changes is always easier than a big diff. :-)

You can also e-mail me other questions (on Git) if you want!

Keep up the good work so far! :+1: :clap:

**Comment by Genki-S** [comment]
> Instead of defining find_by_name, you can just pass self to define_tasks inside TTNT::TestTask#initialize. Not sure if it's useful to keep instances then.

Actually I noticed it shortly after opening this PR and put the fix as this final commit https://github.com/Genki-S/ttnt/commit/3c2cca9679d5b2d8758ac44e5fc1efa9c75d0798
I'm sorry it was confusing, big PR is not good :disappointed: 

> RubyGems monkey-patches require but that only works for released gem

I see, that makes sense. Thank you for the information!

> Actually for now we are only supporting this kind of task but in future we could support other testing frameworks like RSpec.
>  You should rather rely on composition for now

You are right. I am so happy getting your review since I was feeling I was doing a Very Bad Thing :tm: 
Thank you for guiding me to the right direction! :smile: 

> if you want to gather changes, make sure that they are closely related
> try to avoid as much as possible merge commits
> you can have a look at this article for commit messages.

OK, I will check these out and try to make this PR a little cleaner (and will force push). Thank you for your generous advise, I can learn a lot from them :smile: Let me e-mail you when I'm stuck!

> Managing commits may be the less funny part of a patch but if it's done well, it can really make you save some time when fixing bugs or tracking regressions in the future.

Personally I like keeping clean history, so it's no problem! Rather I'm looking forward to make this PR better :+1: I do lots of git-blame when I'm reading code :eyes: 

> Finally, you can create as many pull requests as you wish. Reviewing little changes is always easier than a big diff. :-)

Got it! I learned big diff and dirty history is so confusing... :disappointed: I will take the many little changes path. Thank you for your advice!

---

I will improve the commit history and will let you know when I'm done!

**Comment by Genki-S** [comment]
I finished up my work on this PR. I dropped 2 things from this PR:

- Ability to run tests from within the task (now, currently outputs selected test files)
- Making `TTNT::TestTask` extend `Rake::TestTask`

So, this PR do just these things:

- Regard name parameter passed to `Rake::TestTask` and create tasks named `ttnt:NAME:anchor` and `ttnt:NAME:run`
- Make `TTNT::TestTaks` define tasks rather than making `TTNT::TaskHelper` define them
  - And remove `TTNT::TaskHelper`

I hope my commit messages are clear enough to describe these changes :smile: 

@robin850 Could you review my improvements? I made this PR small, and I will try relying on composition on another PR.

(BTW, I read somewhere that force pushing is not good if others are working on the same branch. I'm thinking it's OK to force push since it's only me committing on this branch. But just in case I saved the reference to previous commits on my local repository so I can restore it if I had better do so)

File: test/testtask_test.rb [hunk 1]
```diff
@@ -3,18 +3,28 @@
...
 class TestTaskTest < Minitest::Test
-  def test_attributes
-    rake_task = nil
+  def setup
+    @name = 'sample_name'
+    @rake_task = nil
     # This will be in users' Rakefiles
-    Rake::TestTask.new { |t|
-      t.libs << "test"
+    @ttnt_task = Rake::TestTask.new { |t|
```
Comment (robin850) [inline test/testtask_test.rb]: I guess this is defined at the wrong place, you should rather do `@ttnt_task = TTNT::TestTask.new(t)` line 14.

**Comment by robin850** [comment]
> BTW, I read somewhere that force pushing is not good if others are working on the same branch.

Yep, this is not really a problem when you are working alone and especially on a pull request. :-)

Looking great, I've just left a tiny comment! The commits look way better ! :+1:

File: test/testtask_test.rb (hunk 1 above)
Comment (Genki-S) [inline test/testtask_test.rb]: Oops, you are right. Thank you for noticing!

Fixed here: https://github.com/Genki-S/ttnt/commit/c660fff10a6020c8d1da7a6b87366874c28b8536

**Comment by Genki-S** [comment]
> Yep, this is not really a problem when you are working alone and especially on a pull request.

Got it, I'll remember it :smile: 

> Looking great, I've just left a tiny comment! The commits look way better !

Thank you so much for your advices! I'll merge this once CI succeeds :+1:

**Pull request merged**